PYIC-6402: Update back event handler

AmritSidhu opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Check previous states until its another page event Manually add back event for SELECT_DEVICE_PAGE as the previous step is a unqiue IDENTIFY_DEVICE where its a page event which doesnt respond with a page

### Why did it change

We can allow users a degree of flexibility to return to previous screens in the journey using a ‘back’ link on the mobile triage page if they want to

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-6402](https://govukverify.atlassian.net/browse/PYIC-6402)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [x] No environment variables or secrets were added or changed


[PYIC-6402]: https://govukverify.atlassian.net/browse/PYIC-6402?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

AmritSidhu committed:
PYIC-6402: Update back event handler

Check previous states until its another page event
Manually add back event for SELECT_DEVICE_PAGE as the previous step is a unqiue IDENTIFY_DEVICE where its a page event which doesnt respond with a page

MikeCollingwood commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -298,7 +298,10 @@ private StepResponse executeJourneyEvent(
-    @SuppressWarnings("java:S3776") // Cognitive Complexity of methods should not be too high
+    @SuppressWarnings({
+        "java:S3776", // Cognitive Complexity of methods should not be too high
+        "java:S6541" // "Brain method"
```
Would it work putting the back event logic into its own method?

MikeCollingwood commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -326,7 +329,23 @@ private State executeStateTransition(
 
-            if (isPageState(initialJourneyState) && isPageState(previousJourneyState)) {
+            // Loop to find the last valid page state
+            while (previousJourneyState != null && !isPageState(previousJourneyState)) {
```
Are we whitelisting all states for skipping over, I thought we were limiting it?

blakeyp commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -396,6 +388,32 @@ private State executeStateTransition(
+            if (previousJourneyState != null
+                    && previousJourneyState
+                            .state()
+                            .equals("STRATEGIC_APP_TRIAGE/IDENTIFY_DEVICE")) {
```
I'm a bit nervous about hardcoding this, what if we rename the state and forget to change it here?
Is there something else we could use to identify this state, perhaps a property on the state itself (eg. `skipBack`) (or this might point to the other idea of defining a bespoke response type for this)

AmritSidhu committed:
PYIC-6402: Add skipBack property to page state

skipBack property is a boolean flag which can be added to page states, if set to true the back functionality will ignore this state and try the previous state
This makes the back functionality more robust a reduces the risk of someone forgetting to change the name state here

AmritSidhu committed:
PYIC-6402: Pull skipBack from pageStepResponse

Add corresponding api-test

MikeCollingwood commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -396,6 +388,28 @@ private State executeStateTransition(
+            if (skipBackResponse != null && Boolean.TRUE.equals(skipBackResponse)) {
+                ipvSessionItem.popState();
+                previousJourneyState = ipvSessionItem.getPreviousState();
+            }
```
I realise this could just check `Boolean.TRUE.equals(skipBackResponse)`, since it would also handle skipBackResponse being null.

MikeCollingwood reviewed (dismissed):
Looks good, API & unit tests are helpful. I think one small refactor is possible but otherwise LGTM

AmritSidhu committed:
PYIC-6402: Remove null check
